### ISSUE TITLE
Using a mmap file instead of normal file io

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#4274]
## Motivation

Currently we using the normal `FileChannelReader` of atomix, which has several problems. It reads a big chunk of data (2 * maxEntrySize, currently per default 4 MB) at once which leads to #4249. Apart of the cache misses there are several open issues #4248 #4252 . 

Atomix also supports memory mapped files and uses here a better approach to chunk and read the data. Memory mapped files have several advantages over normal File io, since the File is mapped into the user space memory there are less copy operations during the read of data. It is much more performant then normal IO. For more information see also this https://howtodoinjava.com/java7/nio/memory-mapped-files-mappedbytebuffer/ and below.

## Investigation

During our engine latency investigation we tried the mmap implementation of atomix and had for example [less cache misses](https://raw.githubusercontent.com/Zelldon/perf-tests/master/mmap/mmap-entry4MB-seg128mb.txt), [lower latency and higher throughput](https://github.com/Zelldon/perf-tests/blob/master/mmap/mmap-results.txt).

## Related

In our [old logstorage](https://github.com/zeebe-io/zeebe/tree/0.20.0/logstreams/src/main/java/io/zeebe/logstreams/impl/log/fs) implementation we also used memory mapped files, you can see [here](https://github.com/zeebe-io/zeebe/blob/0.20.0/logstreams/src/main/java/io/zeebe/logstreams/impl/log/fs/FsLogSegment.java#L77).

## Possible Problems

Due to RAFT we need to make sure that the Leader has written to disk when he commits. In our previous implementation we used a rate limiter to not flush always but periodically.

### Memory Mapped Files

> Possibly the main advantage of Memory Mapped IO is the performance which is important to build high-frequency electronic trading system. Memory Mapped Files are way faster than standard file access via normal IO. Another big advantage of memory mapped IO is that it allows you to load a potentially larger file which is not otherwise accessible. Experiments show that memory mapped IO performs better with large files

> 1. Java supports Memory mapped IO with java.nio package.
> 2. Memory mapped files is used in performance sensitive application, e.g. high-frequency electronic trading platforms.
> 3. By using memory mapped IO, you can load portion of large files in memory.
> 4. Memory mapped file can result in page fault if requested page is not in memory.
> 5. Ability to map a region of the file in memory depends on the addressable size of memory. In a 32 bit machine, you cannot access beyond 4GB or 2^32.
> 6. Memory mapped IO is much faster than Stream IO in Java.
> 7. Memory used to load File is outside of Java heap and resides on shared memory which allows two different processes to access File.
> 8. Reading and writing on memory mapped file is done by the operating system, so even if your Java program crashes after putting content into memory, it will make to File until OS is fine.
> 9. Prefer Direct Byte buffer over NonDirect Buffer for faster performance.
> 10. Don't call MappedByteBuffer.force() method too often, this method is meant to force operating system to write content of memory into disk, So if you call force() method each time you write into memory mapped file, you will not see true benefit of using mapped byte buffer, instead it will be similar to disk IO.
> 11. In case of power failure or host failure, there is a slim chance that content of memory mapped file is not written into the disk, which means you could lose critical data.

https://www.codeproject.com/Tips/683614/Things-to-Know-about-Memory-Mapped-File-in-Java

**Comment by npepinpe** [comment]
I tested to see the gain when enabling this + flushing on every append, with entry size of 256kb, 3 partitions, 3 nodes, replication factor of 3.

1. Append latency went up from 5-10ms average to 15-25ms average, so at least double
1. On the initial start, throughput was worse when one node was leader for all partitions, average ~70wfi/s, with back pressure being the reason for it (e.g. dropping a lot of requests)
1. At some point after some fail overs (intentional), even when one node was leader of all, throughput was better though ~100 wfi/s, so seems back pressure was a little too aggressive there
1. Processing latency was similar, clustered in the 2.5-5s bucket.
1. However, when causing fail overs, and going back to one node leader of all partitions, throughput was up averaging between 140-160 wfi/s, so similar, if slightly better throughput. Latency was comparable though.
1. This then went back down to ~90 wfi/s as back pressure kicked in, though not sure why if I never saw a different in latency.

So from my side it's not clear if we have some immediate gains, since back pressure seems to kick in really aggressively. Also write I/O is very much always ~10-15Mbps, which is meh I guess.

So basically not very promising to flush on every append, though I'll gladly re-do this with someone to verify I was doing this properly :slightly_smiling_face:

**Comment by Zelldon** [comment]
Do you think it makes a difference to flush on commit?

**Comment by npepinpe** [comment]
So your comment made me realize I implemented a very naive and stupid approach :sweat_smile: 

I tested above by flushing on every write, which slowed down the leader by a lot. I switched now to flushing on every append on the follower, and every commit on the leader, which afaik should be safe as well.

You can follow the benchmark on Grafana under `np-mmap`. Not linking since our pw there isn't exactly secure :stuck_out_tongue: 

With 3 nodes, 3 partitions, replication factor 3, flushing on every follower append and flushing on every leader commit, no other special configuration, I get... :drum: :drum: :drum: 

1. Average throughput of 170-180 wfi/s (fluctuates between 150 and 250 wfi/s)

Throughput:

![image](https://user-images.githubusercontent.com/43373/84575196-af488600-adab-11ea-94cd-95c5caee2a15.png)

Latency:

![image](https://user-images.githubusercontent.com/43373/84575182-9e981000-adab-11ea-960c-ce28045bcd13.png)

So actually I would say very promising :smile: 

I didn't even try with lowered message size :wink:

**Comment by npepinpe** [comment]
As mentioned, restarted with 256kb max message size and 128mb segment size, will let it run over the weekend to see if it's less stable the other storage format.

**Comment by npepinpe** [comment]
Pushed a branch, `4274-use-mmap` for posterity

**Comment by npepinpe** [comment]
The benchmark generated a log inconsistency - whether this is due to the new storage format or not is unclear, but considering it reads the log correctly just finds wrong position leads me to think that it's not and we have a bug somewhere that causes inconsistencies!

**Comment by npepinpe** [comment]
Hm, turns out I got a SEGFAULT due to the memory mapped stuff, so maybe the inconsistency is related? Attaching the error log

https://gist.github.com/npepinpe/beae383c09b2906f773bb236e0b2693c

**Comment by npepinpe** [comment]
I think I have an idea - remember how we know readers are probably not thread safe? Well the reader and the writer share the same buffer, but it's managed by the writer, so I'm wondering if that wouldn't cause this issue. Also the seg fault occurred during the inconsistency check, though unsure if that's related.

**Comment by korthout** [comment]
Last week, I've worked on making the usage of memory mapped (mmap) files for the log stream possible in replicated systems. We considered multiple solutions:

1. private buffers: splitting the currently shared mapped buffers between the SegmentWriter and SegmentReaders to private mapped buffers, removing the race conditions between the writer and readers during closing (and concurrently reading);
2. shared buffers: continue with the current shared mapped buffers, but change when buffers are freed, removing the race conditions between the writer and readers during closing (and concurrently reading);
3. private buffer for writing and an additional shared buffer for reading, and changing when buffers are freed.

These solutions require that entries are flushed at the right moments. So the leader must flush on all commits and the followers on all appends, which might lead to more flushing.

Last Friday, I've ran a replicated benchmark (just `make zeebe starter worker`) on the following code bases and configurations:
1. source: develop + allows replicated mmap; config: mmap disabled
2. source: develop (not-threadsafe mmap) + allows replicated mmap; config: mmap enabled
3. source: private buffers threadsafe mmap implementation; config: mmap enabled
4. source: shared buffers threadsafe mmap implementation; config: mmap enabled

The results were the following:
1. throughput: ~150/s; latency: ~3s
2. throughput: ~100/s; latency: 2-3s (but very unstable)
3. throughput: ~155/s; latency: 2-3s: avg ~2.5
4. throughput: ~155/s; latency: 2-3s: avg ~2.5

So, it appears that memory mapped files in a replicated system has an unchanged throughput compared to the current filechannel implementation. However, latency drops a bit (maybe 20% or so). In addition, it shows that these new implementations are a lot more stable than the current mmap implementation.

Considering that private buffers are better maintainable (less complexity due to isolation), it makes sense to continue with that implementation.

**Comment by korthout** [comment]
I've ran some additional benchmark tests to validate the stability as well as the performance improvements made by #5463 

In this comment I will try to provide a comprehensive overview of both the test and the results in such a way that they can also be reproduced by others. At the end, I'll provide a TL;DR. 

## The test setup

All of the following tests are performed using our own benchmark project (see: https://github.com/zeebe-io/zeebe/tree/develop/benchmarks). Zeebe docker images were build for 2 different versions:

- `baseline` (1ecb93164946f64cb9d6f45a7086597abfefb521): only diff to develop was that this lifts the restriction to use memory mapped log segments with a replication faction > 1
- `SNAPSHOT-korthout-4274-private-buffer` (7022bc072302a7ef6ccdb431302b10dacb26d233): contains the full diff of #5463 

> I've updated the commit hashes, because the branch was rebased. The new commit hashes are expected to stay part of the history, so can be used for reproducibility.

The images were build locally. For example, for the image of `SNAPSHOT-korthout-4274-private-buffer`:
```
# builds the dist with zbctl packed
clients/go/cmd/zbctl/build.sh && mvn clean install -T1C -DskipTests -pl dist -am

# builds the a new zeebe docker image
docker build --build-arg DISTBALL='dist/target/zeebe-distribution-*.tar.gz' -t gcr.io/zeebe-io/zeebe:SNAPSHOT-korthout-4274-private-buffer --target app .

# pushes the image to our docker registry
docker push gcr.io/zeebe-io/zeebe:SNAPSHOT-korthout-4274-private-buffer
```

### Test scenarios
Multiple benchmark projects were created using the `newBenchmark.sh` script in `benchmarks/setup`. Each project's `zeebe_values.yaml` file was updated to point to the corresponding docker image, the cpu limit and request were increased to 7 cpus, and the `ZEEBE_BROKER_DATA_USEMMAP` environment variable was altered. The following scenarios were constructed using this:

1. `baseline` with `ZEEBE_BROKER_DATA_USEMMAP`: `false`
2. ~~`baseline` with  `ZEEBE_BROKER_DATA_USEMMAP`: `true`~~ unstable
3. `SNAPSHOT-korthout-4274-private-buffer` with  `ZEEBE_BROKER_DATA_USEMMAP`: `false`
4. `SNAPSHOT-korthout-4274-private-buffer` with `ZEEBE_BROKER_DATA_USEMMAP`: `true`

These scenarios all have a cluster size, replication factor and partition count of 3.

Note that scenario 2 lead to instability, just like in the previous benchmark test (see previous comment), and thus has been removed from the results. Also note that scenario 3 is interesting, because after these changes it will remain the default config in the near future as further confidence in the stability of the memory mapped feature has to be build.

### Config change to the starter
We had a gut-feeling that the benchmark test described in the previous comment was not as successful as we hoped because of the high load we put on zeebe in the standard test. E.g. the starter tries to create 300 process instances per second. So we decreased the starter's rate to 50 (this magic number was only chosen to see if our gut-feeling was correct).

## The test results
The following table shows screenshots of the metrics dashboard of the different benchmark scenarios.

| | 1. baseline no mmap | 3. snapshot no mmap | 4. snapshot mmap | 
|-|-|-|-|
|General|![baseline general](https://user-images.githubusercontent.com/3511026/95751606-27ae4d80-0c9f-11eb-9bed-0fcd13587931.png)|![snapshot no mmap general](https://user-images.githubusercontent.com/3511026/95752452-537e0300-0ca0-11eb-83af-a45f3b683400.png)|![snapshot mmap enabled general](https://user-images.githubusercontent.com/3511026/95753012-3d247700-0ca1-11eb-88c0-cd5cdde5b24b.png)|
|General|![baseline general](https://user-images.githubusercontent.com/3511026/95751614-2bda6b00-0c9f-11eb-9e94-40bfa1a86984.png)|![snapshot no mmap general](https://user-images.githubusercontent.com/3511026/95752471-57aa2080-0ca0-11eb-9453-ad9f2c5a88ce.png)|![snapshot mmap enabled general](https://user-images.githubusercontent.com/3511026/95753023-41509480-0ca1-11eb-8b39-33d9011a5c97.png)|
|Resources|![baseline resources](https://user-images.githubusercontent.com/3511026/95751616-2c730180-0c9f-11eb-987a-c3d9afb476ce.png)|![snapshot no mmap resources](https://user-images.githubusercontent.com/3511026/95752472-58db4d80-0ca0-11eb-9ec7-6948163e0c65.png)|![snapshot mmap enabled resources](https://user-images.githubusercontent.com/3511026/95753026-41e92b00-0ca1-11eb-8c35-4a34eae8bc13.png)|
|**Latency**|![baseline latency](https://user-images.githubusercontent.com/3511026/95750828-fda85b80-0c9d-11eb-8a1c-a910071713b4.png)| ![snapshot no mmap latency](https://user-images.githubusercontent.com/3511026/95751070-58da4e00-0c9e-11eb-87df-1ca6e3506892.png)| ![snapshot mmap enabled latency](https://user-images.githubusercontent.com/3511026/95751191-8cb57380-0c9e-11eb-8ef6-8ba4c0f4b10e.png)|
|**Latency**|![baseline latency](https://user-images.githubusercontent.com/3511026/95750890-19136680-0c9e-11eb-9897-4d7267c8d1ca.png)| ![snapshot no mmap latency](https://user-images.githubusercontent.com/3511026/95751078-5d066b80-0c9e-11eb-8f4b-97e2f9908839.png)| ![snapshot mmap enabled latency](https://user-images.githubusercontent.com/3511026/95751203-90e19100-0c9e-11eb-927a-ca251b5dad25.png)|
|Atomix|![baseline atomix](https://user-images.githubusercontent.com/3511026/95751797-66440800-0c9f-11eb-8689-5f29497bf9fe.png)|![snapshot no mmap atomix](https://user-images.githubusercontent.com/3511026/95752565-7c9e9380-0ca0-11eb-9423-d4392d5cd7d8.png)|![snapshot mmap enabled atomix](https://user-images.githubusercontent.com/3511026/95753153-7826aa80-0ca1-11eb-86c1-9ac315e33a26.png)|
|Atomix|![baseline atomix](https://user-images.githubusercontent.com/3511026/95751815-6b08bc00-0c9f-11eb-8917-a32d89fd7251.png)|![snapshot no mmap atomix](https://user-images.githubusercontent.com/3511026/95752577-80cab100-0ca0-11eb-8a43-d174a944f0de.png)|![snapshot mmap enabled atomix](https://user-images.githubusercontent.com/3511026/95753164-7c52c800-0ca1-11eb-87e7-dafbb76094ba.png)|

These results show a completely different story than the results of the test in the previous comment. Specifically interesting is the latency tab. It shows improvements of the average cycle times with a factor of 3 to 6.5 times (1.3s to 1.5s for non-mmap and 0.2s to 0.5s for mmap enabled). It also shows a clear improvement in processing latency for both commands and events, which is between 25ms and 250ms for non-mmap and between 0ms and 25ms for mmap enabled. Job life time (i.e. the time between when a job was created and when it was completed) has also clearly shrunk from 100-250ms for non-mmap to 50-100ms for mmap enabled.

Also interesting is the resources tab, which shows that mmap enabled leads to a high increase in buffer pool memory usage (from 300MB to 10GB). Since the memory mapping is controlled by the OS, this appears to just be the amount of data that is mapped by us, but not necessarily the memory it takes. We make sure we flush data at the correct moments, and for the rest it's up to the OS to put these files into memory or not.

There's more interesting data in these screenshots, like Atomix flushing and such, but the latency improvement and the resources it costs are the most interesting for this post.

## Stability
We've given both scenario 3 and 4 an additional weekend to run to see the stability of these changes. Both scenarios stayed very stable under our normal operation conditions (our k8s nodes are preemptible which lead to tens of restarts per day) and maintained the throughput (50/s) and latency (~1.5s for non-mmap, 0.2s to 0.5s for mmap enabled).

These changes will be further tested for stability in the upcoming code freeze for release 0.25, and together with that test it gives us the confidence that memory mapped can now be used. Please note, that even with these tests memory mapped log segments are still considered an experimental feature as they are not yet sufficiently tested. These test results only show stability in a relatively short period.

## Non-replicated Zeebe
I've also ran a short benchmark test with the same setup, but the following config:
- `clusterSize`: 3
- `replicationFactor`: 1
- `partitionCount`: 9

This showed similar results in latency as the replicated test.

## Future work
The starter rate of 50 was perhaps not representative of the throughput Zeebe can handle. In long running tests, we currently see an equilibrium at a throughput of ~150/s on our hardware in our google k8s environment, using the benchmark defaults. It might be interesting to see whether we reach the same cycle time improvements if we push zeebe to somewhere between 80-100% of that throughput.

## TL;DR
`ZEEBE_BROKER_DATA_USEMMAP: "true"` seems to stably provide a considerable cycle time improvement (3-6.5x) for both replicated as well as non-replicated zeebe's as long as throughput is not pushed to the limits. Give it a try if you're interested in reducing cycle time (the time it takes 1 process from start to finish), but note that this is still considered an experimental feature.
